File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/F._Максимальная_глубина/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>

using namespace std;

void dfs(const Node* root, int depth, int& maxDepth) {
    if (!root) return;
    ++depth;
    maxDepth = std::max(maxDepth, depth);
    dfs(root->left, depth, maxDepth);
    dfs(root->right, depth, maxDepth);
}


int Solution(const Node* root) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    int maxDepth = 0;
    dfs(root, 0, maxDepth);
    return maxDepth;
}

#ifndef REMOTE_JUDGE
void test() {
    Node node1({1, nullptr, nullptr});
    Node node2({4, nullptr, nullptr});
    Node node3({3, &node1, &node2});
    Node node4({8, nullptr, nullptr});
    Node node5({5, &node3, &node4});
    assert(Solution(&node5) == 3);
}


int main() {
  test();
}
#endif

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/F._Максимальная_глубина/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def solution(root: 'Node') -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(H)
    """
    max_depth: int = 0

    def dfs(root: 'Node', depth: int) -> int:
        if not root: return
        nonlocal max_depth
        depth += 1
        max_depth = max(max_depth, depth)
        dfs(root.left, depth)
        dfs(root.right, depth)

    dfs(root, 0)
    return max_depth


def test():
    node1 = Node(1, None, None)
    node2 = Node(4, None, None)
    node3 = Node(3, node1, node2)
    node4 = Node(8, None, None)
    node5 = Node(5, node3, node4)

    assert solution(node5) == 3


if __name__ == '__main__':
    test()